### PR TITLE
feat(card): Report full screen state as a Shiny input

### LINKED ISSUE #962: [card] Report full screen state as a Shiny input
When `card(full_screen = TRUE, id = "user_provided_id")`, the full screen state could be reported at `input$user_provided_id`.

For future feature compatibility, I'm suggesting that the state be reported as `list(full_screen = TRUE | FALSE)`, so user could would inspect `input$user_provided_id$full_screen` rather than providing a boolean value directly to the input named by the card's id.

```r
library(shiny)
library(bslib)

ui <- page_fluid(
  layout_columns(
    card(
	  id = "my_card",
      full_screen = TRUE,
      uiOutput("my_card_text")
	),
    # ... more cards ....
  )
)

server <- function(input, output, session) {
  output$my_card_text <- renderUI({
    if (input$my_card$full_screen) {
      p("Quick preview. Expand to read more...")
    } else {
      p("Very long, full text that's only available when the card is expanded.")
    }
  })
}

shinyApp(ui, server)
```

### PR DESCRIPTION
**Pull request by gadenbuie**
Closes #962

When `card()` is given an `id`, e.g. `card(id = "my_card")`, the full screen state of the card is now reported as `input$my_card$full_screen`.

Note: this PR builds on #1005 but could be rebased on `main` to decouple.

## Example

```r
library(shiny)
# library(bslib)
pkgload::load_all()

ui <- page_fixed(
  card(
    card_header("Hello card!"),
    "A regular full-screenable card.",
    id = "my_card",
    full_screen = TRUE
  ),
  value_box(
    title = "Hello value box!",
    value = "$1,234,567",
    id = "my_value_box",
    full_screen = TRUE
  ),
  div(
    verbatimTextOutput("debug"),
    style = htmltools::css(
      z_index = 9999,
      position = "absolute",
      bottom = "2rem",
      left = "2rem",
      right = "2rem"
    )
  )
)

server <- function(input, output, session) {
  output$debug <- renderPrint({
    str(
      list(
        my_card = input$my_card,
        my_value_box = input$my_value_box
      )
    )
  })
}

shinyApp(ui, server)
```

### CODE REVIEW

**Comment by cpsievert** [comment]
> Note: this PR builds on https://github.com/rstudio/bslib/pull/1005 but could be rebased on main to decouple.

Please do rebase on main as I have more reservations about that PR.

File: srcts/src/components/card.ts [hunk 1]
```diff
@@ -184,6 +187,25 @@ class Card {
...
+  private _setShinyInput(): void {
+    if (!this.card.classList.contains(Card.attr.CLASS_SHINY_INPUT)) return;
+    if (!Shiny) return;
+    if (!Shiny.setInputValue) {
+      if (Shiny.shinyapp) {
+        Shiny.shinyapp.taskQueue.enqueue(() => this._setShinyInput());
+      } else {
+        setTimeout(() => this._setShinyInput(), 0);
+      }
+      return;
```
Comment (cpsievert) [inline srcts/src/components/card.ts:202]:
```suggestion
```

File: srcts/src/components/card.ts [hunk 2]
```diff
@@ -184,6 +187,25 @@ class Card {
...
+    }
+    Shiny.setInputValue(this.card.id, {
+      // eslint-disable-next-line @typescript-eslint/naming-convention
+      full_screen:
+        this.card.getAttribute(Card.attr.ATTR_FULL_SCREEN) === "true",
+    });
```
Comment (cpsievert) [inline srcts/src/components/card.ts:208]: Pull this logic out into a `_doSetShinyInput` method. That way `_setShinyInput` doesn't need to be recursive

Comment (gadenbuie) [reply srcts/src/components/card.ts:208]: This is intentional and I want `_setShinyInput()` to be recursive. If we have `Shiny` but it doesn't have the `setShinyInput()` method yet, we don't have a reliable non-jQuery way to wait until `Shiny` is fully ready. Neither approach is guaranteed to cause `_setShinyInput()` to run when `Shiny` is fully ready and they may need another cycle or two before we get there.

Happy to add a comment about this in the source.

**Review by cpsievert — APPROVED** [review]

**Pull request merged**